### PR TITLE
[filebeat][azure-blob-storage] - Added support for new features and removed partial save mechanism

### LINKED ISSUE #35126: [Meta] Improve Azure blob storage input
1) Add root level splitting for JSON arrays, similar to GCS input.
2) Add expand event list from field, similar to awss3 input. 
3) ~~Revisit the dynamic cursor implementation and make concurrency improvements similar to the ones made in the gcs 
     input.~~
4) Add file-selector support similar to awss3 input.
5) Add debug logs.
6) Add a date filter for blobs.
7) ~~Decouple object timeout from from scheduler timeout.~~
8) ~~Fix flakey tests.~~ 

Some open issues requesting similar enhancements: 
1) https://github.com/elastic/beats/issues/35186

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Enhancement
- Docs

## Proposed commit message
Added support for the following features as part of a planned enhancement update: 
1) Root level splitting for JSON arrays, similar to GCS input.
2)  File selectors support similar to aws-s3 input, but only supports regex matching atm.
3) A date filter for blobs based on unix epoch (seconds) format.
4) Expand event list from field similar to aws-s3 input.
5) Removed partial save mechanism for now due to concurrency issues.

These additions help make the input more robust for further use. Also added a couple of debug logs for better troubleshooting.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes #35126 #35186

### CODE REVIEW

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:240]: nit
```suggestion
// splitEventList splits the event list into individual events and publishes them.
```

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:248]:
```suggestion
	raw, found := jsonObject[key]
```

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:324]:
```suggestion
// evaluateJSON uses a bufio.NewReader & reader.Peek to evaluate if the
```

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:331]: What is 5?

File: x-pack/filebeat/input/azureblobstorage/mock/data_random.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/mock/data_random.go:107]: Why are we seeding with time? This makes failures non-reproducible, so prefer a defined seed. If you want to have fuzzing, then we should have a Fuzz test.

File: x-pack/filebeat/input/azureblobstorage/mock/testdata/array-at-root.json
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/mock/testdata/array-at-root.json:10]: Final new line?

**Review by efd6 — COMMENTED** [review]
Initial look. I will have a deeper read tomorrow.

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/job.go:168]: yes

**Review by bhapas — COMMENTED** [review]
Added few nits.. 

Overall the unit testing seems to be limited. How is this tested otherwise?

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/job.go:331]: This logic was copied from the gcs input. If I remember correctly, I think this choice for the value **5** was arbitrary, basically tries to read in incremental small chunks. This value worked well for the use case hence never changed it.

File: x-pack/filebeat/input/azureblobstorage/mock/data_random.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/mock/data_random.go:107]: will fix the seed value

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/job.go:157]: all of the rests trigger this method internally

Comment (bhapas) [reply x-pack/filebeat/input/azureblobstorage/job.go:157]: Yeah.. But there are lot of `if`s in this function and probably the straight scenarios are tested. Are all the corner cases covered?

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:249]: Actually partial state saves is a scenario we would ideally want, so this statement acts as a disclaimer sort of, since partial saving is disabled here. Partial saving allows the process to resume from the exact point if the input crashes but can get complex to manage in this scenario, hence not implemented.

Comment (bhapas) [reply x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:249]: Ok.. Probably add it as a Note to highlight that then?

File: x-pack/filebeat/input/azureblobstorage/input_test.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input_test.go:440]: yea should be possible

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/job.go:157]: All of the main logic branches are covered here, it has a of 82% excluding only the error blocks which are manly internal library errors which I'm not sure can be simulated.

File: x-pack/filebeat/input/azureblobstorage/input.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input.go:124]: The date "2106-02-07" is significant in the context of Unix timestamps because it represents the maximum representable date for a 32-bit Unix timestamp. But since we are using 64-bit integers here, i'll update it to ```maxTimestamp := time.Date(3000, time.January, 1, 0, 0, 0, 0, time.UTC).Unix()```

File: x-pack/filebeat/input/azureblobstorage/input_test.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input_test.go:440]: added such a test.

**Review by bhapas — APPROVED** [review]
LGTM. But please wait for @efd6 approval

**Comment by ShourieG** [comment]
@efd6 @bhapas addressed all of the suggestions

File: x-pack/filebeat/input/azureblobstorage/input.go
Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/input.go]: Ahh... `time_t` is specified weakly and I would say is probably most commonly defined from an `int`. So the date that is commonly remarked as the end of time is 2038-01-19.

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/job.go:168]: Can we check that at the site that the value becomes ours?

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/job.go:330]: Let's have a constant with a comment explaining this.

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:34]: The convention for field comments reflects the convention for exported label comments; `// client is an azure blob handle.`

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:91]: Is this safe to release prior to the `processAndPublishData` call? I think it has to be or we have a deadlock when we get to `readJsonAndPublish`. Can you check this?

File: x-pack/filebeat/input/azureblobstorage/scheduler.go
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/scheduler.go:95]: Why is this a wait and not an acquire?

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc
Comment (andrewkroh) [inline x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:287]: > v.Properties.LastModified.Unix() < *s.src.TimeStampEpoch { continue

The description doesn't match up with the code I saw. It seems like it filters blobs that are _older_ than the value.

What is the use case for filtering based on a static value? I've seen inputs offer a relative time based filter when user specifies a duration since _now_ like `ignore_older: 168h` or `since: -168h`.

File: x-pack/filebeat/input/azureblobstorage/input.go
Comment (andrewkroh) [inline x-pack/filebeat/input/azureblobstorage/input.go:129]: What determines the valid range? I can understand validating that the value is non-negative. But why is there an upper range? 

The only reason I can think of for checking the upper value is as a sanity check. It probably does not make sense to configure any value that is in the future.

The bound checking can be added to the struct tag and ucfg will enforce it. Like `validate:"min=1"` for >=1. https://pkg.go.dev/github.com/elastic/go-ucfg#Config.Unpack

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc
Comment (andrewkroh) [inline x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:247]: AWS CloudTrail logs in an Azure Blob? Maybe there's as a common Azure example you can use.

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:287]: Will update the description. The use case is some customers in the past specially asked for the ability to filter out based on a specific date time, hence added this

File: x-pack/filebeat/input/azureblobstorage/input_stateless.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input_stateless.go:82]: So the stateless input is triggered only while testing and till now it did not mimic the real world concurrent execution of multiple containers. This fix allows it to do so.

File: x-pack/filebeat/input/azureblobstorage/scheduler.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/scheduler.go:95]: We acquire the limiter when we do a job, which is defined below. When the scheduling of all the jobs in a current poll interval is done we release the limiter. This statement was previously residing in the schedule method, which did not seem right, since in the gcs input it resides here, in scheduleOnce.

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/job.go:168]: Once we fetch the blob pager, this value becomes localized on the machine.

File: x-pack/filebeat/input/azureblobstorage/input.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input.go:129]: Due to this being an optional pointer value, struct tags validation causes a nil pointer deference. And yes this is used as a sanity check for permissible timestamp values.

File: x-pack/filebeat/input/azureblobstorage/config.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/config.go:18]: This was likely overlooked before, will remove it.

File: x-pack/filebeat/input/azureblobstorage/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/job.go:91]: yes, it causes a deadlock if we don't release. And yea it is safe to release it here since we do not perform any op on this this map until the job is complete and the entry is removed, which is again protected with locks.

File: x-pack/filebeat/input/azureblobstorage/input.go
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input.go:120]: yes the append is unnecessary here, will replace with a simple assignment

File: x-pack/filebeat/input/azureblobstorage/input_stateless.go
Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/input_stateless.go:82]: Can you include those details in the comment.

File: x-pack/filebeat/input/azureblobstorage/scheduler.go
Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/scheduler.go:95]: OK, thanks for clarifying.

**Comment by ShourieG** [comment]
@efd6 updated the PR with the removal of the partial saves

**Pull request merged**
